File: lib/core/utils/geo_discovery/rate_guard.dart
import '../../../config/geo_discovery_config.dart';
import '../telemetry.dart';
import 'geo_types.dart';

class GeoRequestRateGuard {
  final _lastAttempt = <String, DateTime>{};

  Future<void> waitForSlot(String queryFingerprint) async {
    final now = DateTime.now();
    final last = _lastAttempt[queryFingerprint];
    if (last != null) {
      final elapsed = now.difference(last);
      final wait = GeoDiscoveryConfig.minRequestInterval - elapsed;
      if (!wait.isNegative && wait > Duration.zero) {
        telemetry.warn(TelemetryEvent.geoRateLimited, {
          'durationMs': wait.inMilliseconds,
          'queryFingerprint': queryFingerprint,
          'status': 'client_wait',
        });
        await Future<void>.delayed(wait);
      }
    }
    _lastAttempt[queryFingerprint] = DateTime.now();
  }
}

GeoDiscoveryError toGeoDiscoveryError(Object error) {
  if (error is GeoDiscoveryError) return error;
  final message = error.toString();
  if (message.contains('cancel') || message.contains('aborted')) {
    return const GeoDiscoveryError(
      'Geo discovery request was cancelled.',
      code: GeoFailureCode.aborted,
      retryable: false,
    );
  }
  return GeoDiscoveryError(message, code: GeoFailureCode.networkError);
}

File: src/core/geo/rateGuard.ts
import { geoDiscoveryConfig } from "@/config/geoDiscovery";
import { geoTelemetry } from "@/utils/telemetry/geoTelemetry";

const sleep = (ms: number, signal?: AbortSignal) =>
  new Promise<void>((resolve, reject) => {
    if (signal?.aborted) {
      reject(new Error("Aborted"));
      return;
    }

    const timeout = setTimeout(resolve, ms);

    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(timeout);
        reject(new Error("Aborted"));
      },
      { once: true },
    );
  });

export class GeoRequestRateGuard {
  private lastAttemptByFingerprint = new Map<string, number>();

  async waitForSlot(queryFingerprint: string, signal?: AbortSignal) {
    const now = Date.now();
    const lastAttempt = this.lastAttemptByFingerprint.get(queryFingerprint) ?? 0;
    const elapsed = now - lastAttempt;
    const waitMs = geoDiscoveryConfig.rateLimit.minIntervalMs - elapsed;

    if (waitMs > 0) {
      geoTelemetry.warn("geo_rate_limited", {
        durationMs: waitMs,
        queryFingerprint,
        status: "client_wait"
      });
      await sleep(waitMs, signal);
    }

    this.lastAttemptByFingerprint.set(queryFingerprint, Date.now());
  }
}

export const sleepWithAbort = sleep;
